src/object_changeset.hpp

```diff
@@ -0,0 +1,135 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OBJECT_CHANGESET_HPP
+#define REALM_OBJECT_CHANGESET_HPP
+
+#include "collection_notifications.hpp"
+
+#include <realm/keys.hpp>
+#include <realm/util/optional.hpp>
+
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace realm {
+
+class ObjectChangeSet {
+public:
+    using ColKeyType = decltype(realm::ColKey::value);
+    using ObjectKeyType = decltype(realm::ObjKey::value);
+    using ObjectSet = std::unordered_set<ObjectKeyType>;
+    using ObjectMapToColumnSet = std::unordered_map<ObjectKeyType, std::unordered_set<ColKeyType>>;
+
+    ObjectChangeSet() = default;
+    ObjectChangeSet(ObjectChangeSet const&) = default;
+    ObjectChangeSet(ObjectChangeSet&&) = default;
+    ObjectChangeSet& operator=(ObjectChangeSet const&) = default;
+    ObjectChangeSet& operator=(ObjectChangeSet&&) = default;
+
+    void insertions_add(ObjectKeyType obj);
+    void modifications_add(ObjectKeyType obj, ColKeyType col);
+    void deletions_add(ObjectKeyType obj);
+    void clear(size_t old_size);
+
+    bool insertions_remove(ObjectKeyType obj);
+    bool modifications_remove(ObjectKeyType obj);
+    bool deletions_remove(ObjectKeyType obj);
+
+    bool insertions_contains(ObjectKeyType obj) const;
+    bool modifications_contains(ObjectKeyType obj) const;
+    bool deletions_contains(ObjectKeyType obj) const;
+    // if the specified object has not been modified, returns nullptr
+    // if the object has been modified, returns a pointer to the ObjectSet
+    const ObjectSet* get_columns_modified(ObjectKeyType obj) const;
+
+    bool insertions_empty() const noexcept { return m_insertions.empty(); }
+    bool modifications_empty() const noexcept { return m_modifications.empty(); }
+    bool deletions_empty() const noexcept { return m_deletions.empty(); }
+
+    size_t insertions_size() const noexcept { return m_insertions.size(); }
+    size_t modifications_size() const noexcept { return m_modifications.size(); }
+    size_t deletions_size() const noexcept { return m_deletions.size(); }
+
+    bool clear_did_occur() const noexcept
+    {
+        return m_clear_did_occur;
+    }
+    bool empty() const noexcept
+    {
+        return m_deletions.empty() && m_insertions.empty() && m_modifications.empty() && !m_clear_did_occur;
+    }
+
+    void merge(ObjectChangeSet&& other);
+    void verify();
+
+    // provides iterator access to keys in the set (unordered)
+    struct ObjectSetIterable {
```

Comment:
I don't understand the purpose of any of these types. This one and ObjectMapIterable just don't do anything (except for rename size() to count() for some reason). ObjKeyIterator does something, but I don't see how it's useful. The code using it (both here and in realm-js) doesn't obviously benefit from iterating over the wrapper compared to iterating over the map itself.

Reply:
Yeah, it doesn't really do much functionally. I just thought it would be a good idea to provide a level of indirection so that there is a defined interface for iteration as the private containers might eventually end up being something other than the `std::` implementations. It turns out being kinda verbose, but I'm not sure what a good alternative would be. Is removing all of these wrappers, and exposing methods like `modifications_begin()` and `modifications_end()` a better idea?

Reply:
Whatever the final implementation ends up being should still expose the STL container interface (at least for reading), so exposing the unordered_map directly doesn't actually mean that replacing it with a different type of hash map would break any of the code using it.